cocotb/queue.py

```diff
@@ -0,0 +1,204 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from typing import Generic, TypeVar
+import collections
+import heapq
+from asyncio import QueueEmpty, QueueFull
+
+import cocotb
+from cocotb.triggers import Event
+
+T = TypeVar('T')
+
+
+class Queue(Generic[T]):
+    """A queue, useful for coordinating producer and consumer coroutines.
+
+    If *maxsize* is less than or equal to 0, the queue size is infinite. If it
+    is an integer greater than 0, then :meth:`put` will block when the queue
+    reaches *maxsize*, until an item is removed by :meth:`get`.
+    """
+
+    def __init__(self, maxsize: int = 0):
+
+        self._maxsize = maxsize
+
+        self._finished = Event()
+        self._finished.set()
+
+        self._getters = collections.deque()
+        self._putters = collections.deque()
+
+        self._unfinished_tasks = 0
+
+        self._init(maxsize)
+
+    def _init(self, maxsize):
+        self._queue = collections.deque()
+
+    def _put(self, item):
+        self._queue.append(item)
+
+    def _get(self):
+        return self._queue.popleft()
+
+    def _wakeup_next(self, waiters):
+        while waiters:
+            event, task = waiters.popleft()
+            if not task._finished:
+                event.set()
+                break
+
+    def __repr__(self):
+        return '<{} at {:#x} {}>'.format(type(self).__name__, id(self), self._format())
+
+    def __str__(self):
+        return '<{} {}>'.format(type(self).__name__, self._format())
+
+    def __class_getitem__(cls, type):
+        return cls
+
+    def _format(self):
+        result = 'maxsize={}'.format(repr(self._maxsize))
+        if getattr(self, '_queue', None):
+            result += ' _queue={}'.format(repr(list(self._queue)))
```

Comment:
I'm not entirely sure how this line isn't being covered. The base implementation has a `_queue` attribute.

Reply:
I think it's because the `_queue` is empty at that point in the test, so the `if` is `False`.